### PR TITLE
#187419176 Seller can avail and unavail product to a buyer

### PR DESCRIPTION
**Pull request by MugemaneBertin2001**
# What does this PR do?
- It enables  a seller to avail an unavail products to a buyers.

# Description of performed task
- Products should be availed for buyers to purchase
by default, the product should be unavailable on creation

- As a seller I should be able to mark a product as available or unavailable manually in order to make it available for the buyers or not.

- Getall / Getone should only list available products
but for the seller, all of his/her products should be listed regardless of the status

## How to test it
1. clone this repo
2. switch to `ft-paroduct-avail-#187419176` branch
3. browse to localhost:port/docs

relevant pivotaltracker story #187419179

### CODE REVIEW

File: src/controllers/productControllers.ts [hunk 1]
```diff
@@ -177,3 +185,12 @@ export const searchProductController = async (req: Request, res: Response) => {
         res.status(500).json({ error: 'Internal server Error' });
     }
 };
+export const productAvailability = async (req: Request, res: Response) => {
+    try {
+        const availability = await updateProductAvailability(req,res);
+        return availability;
+    } catch (error) {
+        console.error('Error fetching product availability:', error);
```
Comment (teerenzo) [inline src/controllers/productControllers.ts]: remove this console

File: src/middlewares/sellerAuth.ts [hunk 2]
```diff
@@ -17,3 +19,43 @@ export const isAseller  = async(req:Request,res:Response,next:NextFunction) => {
...
+    } else {
+      //@ts-ignore
+      req.userRole = loggedUser.role;
+      //@ts-ignore
+      req.userId = loggedUser.id;
+    }
+    next();
+  } catch (error: any) {
+    return res.status(401).json({
+      status: "failed",
```
Comment (teerenzo) [inline src/middlewares/sellerAuth.ts]: where are you using this?

File: src/controllers/productControllers.ts [hunk 3]
```diff
@@ -6,26 +6,34 @@ import {
...
+        //         status:200,
+        //         message:"Products list is empty"
+        //     });
+        // }else{
+        //     res.status(200).json({
+        //         message: "Products fetched successfully",
+        //         count: products.length,
+        //         products: products,
+        //       });
+        // }
```
Comment (teerenzo) [inline src/controllers/productControllers.ts]: why commenting this?

**Review by teerenzo — CHANGES_REQUESTED** [review]

File: src/controllers/productControllers.ts [hunk 4]
```diff
@@ -6,26 +6,21 @@ import {
...
+    searchProduct,
+    updateProductAvailability
+    
 } from "../services/product.service";
 import { ProductType } from "../types";
 
 export const fetchProducts =  async(req:Request,res:Response) =>{
     try {
+        //@ts-ignore
+        console.log(req.viewer)
```
Comment (teerenzo) [inline src/controllers/productControllers.ts:18]: remove this

**Review by teerenzo — CHANGES_REQUESTED** [review]

**Comment by teerenzo** [comment]
work on your PR title, see https://github.com/atlp-rwanda/engineering-playbook/wiki/3.-Pull-Request

**Review by Heisjabo — COMMENTED** [review]
a seller can not see his/her unavalable products

**Review by Jovz19200 — COMMENTED** [review]
Screenshot 2024-05-10 at 14 30 03
Every thing is working well, except when trying to get an unavailable product  which was disabled, I am getting the status code 200 , with an empty product array.
 Which I don't know if it is a good practice, I would suppose a status code 404 Not found instead. but everything else I checked is OK!

**Review by Heisjabo — APPROVED** [review]

**Review by teerenzo — APPROVED** [review]

**Review by Jovz19200 — APPROVED** [review]

**Pull request merged**